Add `makeself` backend

grihabor opened the pull request:
[Makeself](https://github.com/megastep/makeself) is a small shell script that generates a self-extractable compressed tar archive from a directory. This backend provides a `makeself_archive` target which can package a makeself archive. Here is an example:

**src/shell/run.sh:**
```bash
#!/bin/bash
src/shell/hello.sh
src/shell/world.sh
```
**src/shell/hello.sh:**
```bash
#!/bin/bash
printf hello
```
**src/shell/world.sh:**
```bash
#!/bin/bash
printf world
```
**src/shell/BUILD:**
```python
shell_sources()
```
**BUILD:**
```python
makeself_archive(
    name="hello",
    startup_script="src/shell/run.sh",
    files=["src/shell/hello.sh", "src/shell/world.sh"],
)
```
Then you can run `pants package :hello` and it will produce an executable `dist/hello.run` (the file extension is recommended by [makeself maintainer](https://github.com/megastep/makeself)). Here is an [example repo](https://github.com/grihabor/pants-playground/tree/9c6411ea7d313de6fe42063b5444991c18a9ee13/makeself).

This is a minimal implementation and does not yet support all makeself options.

huonw commented on src/python/pants/backend/makeself/goals/package_test.py:
Minor, but given this is actually building archives and examining/executing them, I think the convention would be for this to be called `package_integration_test.py`.

Also, since it seems to be running them too, potentially it could be `package_run_integration_test.py`? Just thinking for someone editing `run` in future and then attempting to find where to add tests. (Either that, or split it up to have `run_integration_test.py` too.)

huonw replied:
Ah, yes, you're right. The pants repo is inconsistent.

For future, Python backends are potentially better reference points, because they're more 'critical'/more maintained, e.g.

- https://github.com/pantsbuild/pants/blob/d0c0fca0c75507a6c2cc783d18a70107174927dc/src/python/pants/backend/python/goals/package_pex_binary_integration_test.py
- https://github.com/pantsbuild/pants/blob/d0c0fca0c75507a6c2cc783d18a70107174927dc/src/python/pants/backend/python/goals/run_pex_binary_integration_test.py

huonw commented on src/python/pants/backend/makeself/target_types.py:
```diff
@@ -0,0 +1,100 @@
+
+class MakeselfArchiveTarget(Target):
+    alias = "makeself_archive"
+    core_fields = (
```
One thing we're considering for tools like PEX is a field for passing through arbitrary extra args so that users can do ~whatever the underlying tool can, even if pants doesn't have explicit support for it (e.g. #20227).

It doesn't have to happen in this PR, but maybe that's an approach that could help here?

huonw commented on src/python/pants/backend/makeself/goals/BUILD:
```diff
@@ -0,0 +1,8 @@
+
+python_tests(
+    name="tests",
+    # dependencies=["pants-plugins/pants.backend.makeself#pantsbuild.pants"],
```
Is this left over?

huonw commented on src/python/pants/backend/makeself/makeself.py:
```diff
@@ -0,0 +1,322 @@
+            rationale="create makeself archive",
+        ),
+    )
+    tooldir = "__makeself"
```
Could this be either moved to a global or placed in `MakeselfTool` (potentially have the `exe` field be the full path, including the `__makeself`?), so that it's not duplicating the `out = ...` in `extract_makeself_distribution`?

huonw commented on src/python/pants/backend/makeself/makeself.py:
```diff
@@ -0,0 +1,322 @@
+
+    if output_directory := request.output_directory:
+        output_directories = [output_directory]
+        argv.extend(["--keep", "--target", request.output_directory])
```
Is this to allow a performance optimisation of only extracting once? Should this also pass `--noexec`?

huonw replied:
Right. So, to confirm, you're saying all callers of this who pass `output_directory` definitely want the start-up script to be run?

huonw commented on src/python/pants/backend/makeself/makeself.py:
```diff
@@ -0,0 +1,322 @@
+        request.exe,
+        "--accept",
+        "--noprogress",
+        "--nox11",
+        "--nochown",
```
Will these cause issues with some `makeself` archives?

huonw reviewed (commented):
Thanks for this.

There's a few other places that also need changes when a new backend is added (very non-obvious)!

- make sure it's packaged into the published artifact: https://github.com/pantsbuild/pants/blob/7a1cd54d98a43485027d2ac53099565900659811/src/python/pants/bin/BUILD
- the listing of all backends: https://github.com/pantsbuild/pants/blob/7a1cd54d98a43485027d2ac53099565900659811/docs/markdown/Using%20Pants/concepts/enabling-backends.md#available-experimental-backends
- might as well add it to the reference docs too: https://github.com/pantsbuild/pants/blob/7a1cd54d98a43485027d2ac53099565900659811/build-support/bin/generate_docs.py#L173-L228

grihabor reviewed (commented):
Thank you for review!

Ready for the second iteration

huonw commented on src/python/pants/backend/makeself/makeself.py:
```diff
@@ -0,0 +1,322 @@
+    test: TestBinary,
+    wc: WcBinary,
+    xz: XzBinary,
+    zstd: ZstdBinary,
```
Okay, I'm happy enough to leave it for a future PR, but it'd be good to set a direction before you start working on it. Given it sounds like you have an idea in mind, could you sketch out the `BUILD` syntax you're thinking of?

huonw replied:
Ah, okay. Thanks for explaining.

Our experience from #19643 is strict validation of input parameters to tools that pants is managing can be annoying. For example, do a manual upgrade of the tool using `pants.toml` to have some new functionality (in that case, a new version of pip; in this case, maybe some new compression or new hashing algorithm), and then not be able use it via pants, because pants is validating the arg too strictly.

So, a version that's more resilient to improvements to `makeself` might allow a bit more customisation without needing new pants' versions. Some possible inspiration would be `shell_command` (with its `tools` argument) and `adhoc_tool` (with its `executable_dependencies` & `system_binary`).

Thoughts?

huonw commented on src/python/pants/backend/experimental/makeself/BUILD:
```diff
@@ -0,0 +1,11 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# NOTE: Sources restricted from the default for python_sources due to conflict with
```
Hm, other `BUILD` files don't seem to need to do this; why is this one special? (Example I checked: https://github.com/pantsbuild/pants/blob/db14a1095d6b0ed4a36b0bf84ee2cbd07085da07/src/python/pants/backend/experimental/adhoc/BUILD)

huonw commented on src/python/pants/backend/experimental/makeself/register.py:
```diff
@@ -1,10 +1,13 @@
-from . import makeself
-from .goals import package, run
-from .target_types import MakeselfArchiveTarget
+See https://github.com/megastep/subsystem.
```
```suggestion
See https://github.com/megastep/makeself.
```

huonw commented on src/python/pants/backend/makeself/subsystem.py:
```diff
@@ -0,0 +1,92 @@
+    help = "A tool to generate a self-extractable compressed tar archives."
+
+    default_version = "2.5.0"
+    default_known_versions = [
+        "2.5.0|macos_arm64 |4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+        "2.5.0|macos_x86_64|4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+        "2.5.0|linux_arm64 |4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+        "2.5.0|linux_x86_64|4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+    ]
```
No particular need to change, but given this is the same for all platforms, feel free to reduce the repetition, similar to what PEX does: https://github.com/pantsbuild/pants/blob/db14a1095d6b0ed4a36b0bf84ee2cbd07085da07/src/python/pants/backend/python/util_rules/pex_cli.py#L42-L54

huonw reviewed (commented):
Thanks for iterating; a few more notes, and then we're probably good to go, as long as if you're going to improve the external tool behaviour as a follow-up 😄

huonw commented:
FYI @grihabor, I'm going to be on leave for 2 weeks from tomorrow. Please feel free to work through my remaining comments and/or solicit more reviews on #development on slack. Thanks for taking the time to contribute!

grihabor commented:
Ok, thank you!

kaos commented on src/python/pants/backend/makeself/goals/run.py:
```diff
@@ -0,0 +1,185 @@
+    pwd: PwdBinary,
+    rm: RmBinary,
+    sed: SedBinary,
+    shasum: DateBinary,
```
??

```suggestion
    shasum: ShasumBinary,
```

grihabor commented:
Rebase changes: removed docs related stuff because it was moved
```
- build-support/bin/generate_docs.py                 |   3 +-
- .../Using Pants/concepts/enabling-backends.md      |   1 +
```

grihabor commented:
@huonw @kaos What do you think? Shall we merge this?

kaos commented on src/python/pants/backend/makeself/goals/package.py:
```diff
@@ -0,0 +1,246 @@
+        request.archive_dir,
+        request.file_name,
+        request.label,
+        os.path.join(os.curdir, request.startup_script),
```
the `os.curdir` here makes me wary.. if using a remote build server, this is likely going to break, as the project repo won't be available then, only the sandbox. Does it work if you change it to `"{chroot}"` if makeself doesn't work with relative paths? (the startup_script would have to be included in the input_digest as well, naturally.)

kaos reviewed (commented):
Apologies for the slow review. I have one remaining comment only.

huonw commented:
Thanks for waiting. I think this is fine to land as experimental (as it is), although I'd like to see some more thinking on https://github.com/pantsbuild/pants/pull/20360#discussion_r1439897543, because I don't think requiring all binaries to exist is a viable path forward for any sort of usage, beyond initial experimentation. 

(There's a chance the new tests here will break someone running `pants test ::` if they don't have the right local setup... but running that should be fairly rare, I think.)

grihabor commented:
@kaos @huonw Thank you for your review!

It turned out that makeself supports not just a single shell script, but arbitrary command with arguments. So I changed the API and I no longer need `os.curdir`.

The changes (https://github.com/pantsbuild/pants/pull/20360/commits/687e1685d3da22ed31485fd64edf04c862cb3353):
- added `makeself_archive.args` for arbitrary makeself args
- `makeself_archive.startup_script` is now a list of tuples, for example, `["./my_script.sh"]` or `["echo", "pants"]`
- added more tests

grihabor commented:
P.S. ready for the next review round.

huonw commented on src/python/pants/backend/makeself/goals/run.py:
```diff
@@ -0,0 +1,191 @@
+@rule(desc="Run makeself archive", level=LogLevel.DEBUG)
+async def run_makeself_archive(
+    request: RunMakeselfArchive,
+    awk: AwkBinary,
+    base64: Base64Binary,
+    basename: BasenameBinary,
+    bash: BashBinary,
+    bzip2: Bzip2Binary,
+    cat: CatBinary,
+    cut: CutBinary,
+    date: DateBinary,
+    dd: DdBinary,
+    df: DfBinary,
+    dirname: DirnameBinary,
+    expr: ExprBinary,
+    find: FindBinary,
+    gpg: GpgBinary,
+    gzip: GzipBinary,
+    head: HeadBinary,
+    id: IdBinary,
+    md5sum: Md5sumBinary,
+    mkdir: MkdirBinary,
+    pwd: PwdBinary,
+    rm: RmBinary,
+    sed: SedBinary,
+    shasum: ShasumBinary,
+    tail: TailBinary,
+    tar: TarBinary,
+    test: TestBinary,
+    wc: WcBinary,
+    xz: XzBinary,
+    zstd: ZstdBinary,
```
I think this long list of required binaries is still problematic (e.g. `gpg` might not be installed... and I imagine people might be uninstalling `xz` recently!), and we don't seem to have wrapped up the conversation in https://github.com/pantsbuild/pants/pull/20360 . I'd like to at least agree on an approximate path forward before we land this.

grihabor replied:
I think my comment got lost somewhere.

Yeah, I'm totally ok to remove the optional binaries and use mechanism from `adhoc_tool` in the next pr.

huonw commented on src/python/pants/backend/makeself/goals/package_test.py:
I think the suggestion to rename this file to `package_run_integration_test.py` from https://github.com/pantsbuild/pants/pull/20360#discussion_r1439937456 would still be good to apply. (And/or split to have a `run_integration_test.py` too.)

grihabor replied:
Sure https://github.com/pantsbuild/pants/pull/20360/commits/b14ce842c6b42ae7a932dcb0864642ab16322f73

huonw commented on src/python/pants/backend/makeself/target_types.py:
```diff
@@ -0,0 +1,102 @@
+    required = False
+    help = help_text(
+        """
+        The startup script, i.e. what gets run when executing `./my_archive.run`.
```
Since this is now a sequence of tuples, presumably to allow passing arbitrary arguments, can we add a note or example about that to this docstring?

grihabor replied:
Sure https://github.com/pantsbuild/pants/pull/20360/commits/3a39ee9d045293993570e787cc6ec3912ed21540

huonw reviewed (commented):
Thanks

grihabor commented on src/python/pants/backend/makeself/target_types.py:
```diff
@@ -0,0 +1,102 @@
+    required = False
+    help = help_text(
+        """
+        Makeself script args, see docs [here](https://github.com/megastep/makeself/tree/release-2.5.0#usage).
```
Thanks! That's a good point, here is the fix https://github.com/pantsbuild/pants/pull/20360/commits/5c5197b893514d86e14737adf0c589756d662372

> Can we find another way to do that?
 
Yeah the user can change the version, so I changed the link to point to README from default branch

grihabor reviewed (commented):
Thank you @huonw!

grihabor commented:
@huonw Shall we merge this one?

huonw commented:
Thanks for waiting.

I tried running these tests locally, without `zstd` available, and observed that tests fail: 

```
pants.core.util_rules.system_binaries.BinaryNotFoundError: Cannot find `zstd` on `['/bin', '/opt/homebrew/bin', '/usr/bin', '/usr/local/bin']`. Please ensure that it is installed so that Pants can zstd file.
```

So, I think we probably can't land this directly, or else we may be breaking running Pants tests on some platforms. Some options are:

1. ensure that the long list of system binaries are optional before landing
2. require that `zstd`, `gpg`, ... are all installed
3. something else?

In addition, once we've got a direction here, you'll need to merge `main` and add some release notes to `docs/notes/2.22.x.md`. See https://github.com/pantsbuild/pants/discussions/20888 for more info.

grihabor commented:
@huonw Thanks for testing! I've reduced the list of binaries to only required ones and added the `tools` field to add more binaries if needed. Also added notes and a documentation page near the shell docs.

huonw commented on docs/docs/shell/self-extractable-archives.mdx:
```diff
@@ -0,0 +1,265 @@
+
+## `pants run`
+
+Instead of packaging and running `makeself_archive` manually, you can use `run` goal instead:
```
```suggestion
Instead of packaging and running `makeself_archive` manually, you can use the `run` goal instead:
```

huonw commented on docs/docs/shell/self-extractable-archives.mdx:
```diff
@@ -0,0 +1,265 @@
+    files=["lib.sh:sh", "entrypoint.sh:sh"],
+    packages=[":upper"],
+    startup_script=["./entrypoint.sh"],
+    tools=["python3.12", "grep", "sort"],
```
Hm, just observing that these tools seem to be used for running... but I imagine they're not required for packaging? Potentially something to iterate on?

grihabor replied:
This is specifically about the pex inside. Since the `run` goal doesn't know there is a pex inside a makeself archive, you need to explicitly add python, grep and sort which are used inside the pex launcher, but not required for makeself launcher.

This is not needed for shell-only makeself archives.

huonw replied:
Yeah, I understand it's just for this particular archive... but I think the tools are only used when running the makeself archive using `pants run`? That is, even if this had no tools specified `tools=[]`, I think the `pants package ...` goal would still work.

No need to change, but just observing that there's a distinction between "tools required to package" and "tools required to run", and overlaying them may lead to spurious cache invalidations (etc.).

grihabor replied:
Ah, yes, this is correct

I could split it into `execution_dependencies` and `runnable_dependencies` similar to `adhoc_tool`

huonw commented on docs/docs/shell/self-extractable-archives.mdx:
````diff
@@ -0,0 +1,265 @@
+```
+
+To pass the arguments through the `pants run` goal you need `--`, then you need
+another `--` to pass arguments to archive `startup_script`, so you end up with
````
```suggestion
another `--` to pass arguments to the archive's `startup_script`, so you end up with
```

huonw commented on docs/docs/shell/self-extractable-archives.mdx:
````diff
@@ -0,0 +1,265 @@
+one
+```
+
+To pass the arguments to the startup_script use `--`:
````
```suggestion
To pass the arguments to the `startup_script` use `--`:
```

huonw commented on docs/notes/2.22.x.md:
```diff
@@ -30,6 +30,15 @@ A new experimental `SQL` backend was added along with the [sqlfluff
 
+#### NEW: Makeself
+
+A new experimental backend was added to support
```
```suggestion
A new experimental `pants.backend.experimental.makeself` backend was added to support
```

huonw commented on docs/notes/2.22.x.md:
```diff
@@ -30,6 +30,15 @@ A new experimental `SQL` backend was added along with the [sqlfluff
+[`makeself_archive`](https://www.pantsbuild.org/2.22/reference/targets/makeself_archive)
+target allows you to create self-extractable archives that bundle files, bash
+scripts or even pex binaries into a single executable archive! See examples in the
+[docs](https://www.pantsbuild.org/2.21/docs/shell/self-extractable-archives).
```
```suggestion
[docs](https://www.pantsbuild.org/2.22/docs/shell/self-extractable-archives).
```

huonw commented on docs/docs/shell/self-extractable-archives.mdx:
Just observing that this is scoped within the "shell" docs, which is nice. Do you think the whole backend could potentially be scoped within the shell one? Or is that unnecessary? Like `pants.backend.experimental.shell.makeself`

grihabor replied:
I believe with some hacks it can be used without shell scripts, e.g. python or any other scripting language that supports shebangs. Most probably you won't do this, but I think we can keep it in it's own space. 

We can come back to this question once we decide to stabilize the backend

huonw reviewed (approved):
Looks great, other than some copy-editing of the docs and a question about the location of the backend!

huonw commented:
(Just being explicit: approval conditional on the copy-editing and answering that question, in case that wasn't clear. 😄 )
